File: OsmAnd-java/src/test/java/net/osmand/shared/compat/SearchWordsBenchmarkTest.java
package net.osmand.shared.compat;

import net.osmand.binary.CommonWords;
import net.osmand.map.OsmandRegions;
import net.osmand.util.LocationParser;

import org.junit.Assume;
import org.junit.Test;

import java.io.File;
import java.io.IOException;
import java.lang.reflect.Constructor;
import java.lang.reflect.Method;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.util.ArrayList;
import java.util.List;

/**
 * The java half of {@code SearchWordsBenchmarkTest} in OsmAnd-shared: how long {@link CommonWords}
 * takes to build its words with those of the regions, which the search does once, before its first
 * phrase; to rank words by them; and how long {@link LocationParser} takes to read places typed as
 * coordinates. The words and phrases are the ones {@link CommonWordsCompatTest} and
 * {@link LocationParserCompatTest} write to their dumps, which this reads.
 *
 * <b>Not part of a normal run</b>: it only does anything when {@code OSMAND_SEARCH_WORDS_BENCHMARK}
 * is set, and measures java alone, in a jvm of its own; the copy is measured by the shared half.
 * <pre>
 * OSMAND_SEARCH_WORDS_BENCHMARK=1 ./gradlew :OsmAnd-java:test --tests "*SearchWordsBenchmarkTest" --rerun -i
 * </pre>
 * On the {@code regions.ocbf} of 24.09.2026 and master of 26.09.2026, best of five in ms; the copy
 * and native columns are the shared half, on the jvm and on the release simulator binary:
 * <pre>
 * what                               java    copy  native   count
 * first build in the process         57.0    43.4    28.6       1
 * build the words                    13.6    12.7    28.8       1
 * rank 91170 words                    2.1     1.9     6.1   51655
 * parse 78372 phrases               139.8   128.9  1085.9   54609
 * </pre>
 */
public class SearchWordsBenchmarkTest {

	private static final File REGIONS = new File("src/main/resources/net/osmand/map/regions.ocbf");
	private static final File WORDS = new File("build/common-words-java.txt");
	private static final File PHRASES = new File("build/location-parser-java.txt");
	private static final int WARMUP_ROUNDS = 3;
	private static final int MEASURED_ROUNDS = 5;

	/** The steps of {@code CommonWords.getInstance}, in its order. */
	private static final String[] BUILD = {
			"addCalculatedAddrCommonWords", "addCalculatedPoiCommonWords", "addAbbrevationsToCommon",
			"addManualAbbrevationsToFrequent", "addRegionNames", "addCalculatedAddrFrequentWords",
			"addCalculatedPoiFrequentWords",
	};

	@Test
	public void benchmarkSearchWords() throws Exception {
		Assume.assumeTrue("set OSMAND_SEARCH_WORDS_BENCHMARK to run", System.getenv("OSMAND_SEARCH_WORDS_BENCHMARK") != null);
		Assume.assumeTrue("run CommonWordsCompatTest and LocationParserCompatTest first", WORDS.exists() && PHRASES.exists());
		List<String> words = inputs(WORDS, "W ");
		List<String> phrases = inputs(PHRASES, "L ");
		net.osmand.PlatformUtil.setOsmandRegions(new OsmandRegions(REGIONS.getPath()));

		double[] best = {Double.MAX_VALUE, Double.MAX_VALUE, Double.MAX_VALUE};
		double first = 0;
		int ranked = 0;
		int places = 0;
		for (int round = 0; round < WARMUP_ROUNDS + MEASURED_ROUNDS; round++) {
			long started = System.nanoTime();
			CommonWords cw = build();
			double build = (System.nanoTime() - started) / 1e6;

			started = System.nanoTime();
			int r = 0;
			for (String w : words) {
				if (cw.getCommonSearch(w) != -1) {
					r++;
				}
			}
			double rank = (System.nanoTime() - started) / 1e6;

			started = System.nanoTime();
			int p = 0;
			for (String phrase : phrases) {
				if (LocationParser.parseLocation(phrase) != null) {
					p++;
				}
			}
			double parse = (System.nanoTime() - started) / 1e6;

			if (round == 0) {
				first = build;
			}
			if (round >= WARMUP_ROUNDS) {
				double[] times = {build, rank, parse};
				for (int i = 0; i < times.length; i++) {
					best[i] = Math.min(best[i], times[i]);
				}
				ranked = r;
				places = p;
			}
		}
		System.out.println();
		System.out.println("### search words, java, best of " + MEASURED_ROUNDS);
		row("first build in the process", first, 1);
		row("build the words", best[0], 1);
		row("rank " + words.size() + " words", best[1], ranked);
		row("parse " + phrases.size() + " phrases", best[2], places);
	}

	/** A new {@link CommonWords}, built as {@code getInstance} builds its own. */
	private static CommonWords build() throws ReflectiveOperationException {
		Constructor<CommonWords> constructor = CommonWords.class.getDeclaredConstructor();
		constructor.setAccessible(true);
		CommonWords cw = constructor.newInstance();
		Method addCommon = CommonWords.class.getDeclaredMethod("addCommon", String.class);
		addCommon.setAccessible(true);
		addCommon.invoke(cw, "NUMBER_WITH_LESS_THAN_2_LETTERS");
		for (String step : BUILD) {
			Method m = CommonWords.class.getDeclaredMethod(step);
			m.setAccessible(true);
			m.invoke(cw);
		}
		return cw;
	}

	private static List<String> inputs(File dump, String prefix) throws IOException {
		List<String> result = new ArrayList<>();
		for (String line : Files.readAllLines(dump.toPath(), StandardCharsets.UTF_8)) {
			if (line.startsWith(prefix)) {
				String h = line.substring(prefix.length() + 1, line.indexOf('\t'));
				StringBuilder sb = new StringBuilder();
				for (int i = 0; i < h.length(); i += 4) {
					sb.append((char) Integer.parseInt(h.substring(i, i + 4), 16));
				}
				result.add(sb.toString());
			}
		}
		return result;
	}

	private static void row(String what, double ms, int count) {
		System.out.println(String.format(java.util.Locale.US, "  %-34s %9.1f %9d", what, ms, count));
	}
}

File: OsmAnd-shared/src/commonTest/kotlin/net/osmand/shared/binary/SearchWordsBenchmarkTest.kt
package net.osmand.shared.binary

import net.osmand.shared.map.OsmandRegions
import net.osmand.shared.routing.testEnvironment
import net.osmand.shared.routing.testPlatformName
import net.osmand.shared.util.KLocationParser
import net.osmand.shared.util.dumpUnhex
import net.osmand.shared.util.readJavaDump
import kotlin.test.Test
import kotlin.time.TimeSource

/**
 * How long the shared copy takes to build the common and frequent words with those of the regions,
 * which the search does once, before its first phrase; to rank words by them; and to read places
 * typed as coordinates. The java half is `SearchWordsBenchmarkTest` in OsmAnd-java, over the same
 * words and phrases: the ones `CommonWordsCompatTest` and `LocationParserCompatTest` write to their
 * dumps, which this reads.
 *
 * **Not part of a normal run**: it only does anything when `OSMAND_SEARCH_WORDS_BENCHMARK` is set.
 * On Kotlin/Native only the **release** binary says anything about speed:
 * ```
 * ./gradlew :OsmAnd-shared:linkReleaseTestIosSimulatorArm64
 * SIMCTL_CHILD_OSMAND_SEARCH_WORDS_BENCHMARK=1 SIMCTL_CHILD_OSMAND_REGIONS_OCBF=<regions.ocbf> \
 *   SIMCTL_CHILD_OSMAND_COMMON_WORDS_JAVA_DUMP=<dump> SIMCTL_CHILD_OSMAND_LOCATION_PARSER_JAVA_DUMP=<dump> \
 *   xcrun simctl spawn --standalone <udid> build/bin/iosSimulatorArm64/releaseTest/test.kexe \
 *   --ktest_filter='net.osmand.shared.binary.SearchWordsBenchmarkTest.*'
 * ```
 *
 * On the `regions.ocbf` of 24.09.2026 and master of 26.09.2026, best of five in ms, next to the
 * java half:
 * ```
 * what                               java    copy  native   count
 * first build in the process         57.0    43.4    28.6       1
 * build the words                    13.6    12.7    28.8       1
 * rank 91170 words                    2.1     1.9     6.1   51655
 * parse 78372 phrases               139.8   128.9  1085.9   54609
 * ```
 * Kotlin/Native builds the words in 29 ms, 2.3 times the jvm once it is warm and 0.66 of a jvm's
 * first time. It ranks a word in 67 ns and reads a phrase in 14 us, 8.4 times the copy on the jvm;
 * the search reads one phrase per query.
 */
class SearchWordsBenchmarkTest {

	@Test
	fun benchmarkSearchWords() {
		if (testEnvironment("OSMAND_SEARCH_WORDS_BENCHMARK") == null) {
			println("SearchWordsBenchmarkTest: set OSMAND_SEARCH_WORDS_BENCHMARK to run")
			return
		}
		val wordLines = readJavaDump(
			"SearchWordsBenchmarkTest", "OSMAND_COMMON_WORDS_JAVA_DUMP", "common-words-java.txt", "CommonWordsCompatTest"
		) ?: return
		val phraseLines = readJavaDump(
			"SearchWordsBenchmarkTest", "OSMAND_LOCATION_PARSER_JAVA_DUMP", "location-parser-java.txt",
			"LocationParserCompatTest"
		) ?: return
		val words = wordLines.filter { it.startsWith("W ") }.map { dumpUnhex(it.substring(2, it.indexOf('\t'))) }
		val phrases = phraseLines.filter { it.startsWith("L ") }.map { dumpUnhex(it.substring(2, it.indexOf('\t'))) }
		val path = testEnvironment("OSMAND_REGIONS_OCBF") ?: "../OsmAnd-java/src/main/resources/net/osmand/map/regions.ocbf"
		val regions = OsmandRegions(path)

		val best = DoubleArray(3) { Double.MAX_VALUE }
		var first = 0.0
		var ranked = 0
		var places = 0
		repeat(WARMUP_ROUNDS + MEASURED_ROUNDS) { round ->
			var mark = TimeSource.Monotonic.markNow()
			val cw = CommonWords.create(regions)
			val build = mark.elapsedNow().inWholeMicroseconds / 1000.0

			mark = TimeSource.Monotonic.markNow()
			var r = 0
			for (w in words) {
				if (cw.getCommonSearch(w) != -1) {
					r++
				}
			}
			val rank = mark.elapsedNow().inWholeMicroseconds / 1000.0

			mark = TimeSource.Monotonic.markNow()
			var p = 0
			for (phrase in phrases) {
				if (KLocationParser.parseLocation(phrase) != null) {
					p++
				}
			}
			val parse = mark.elapsedNow().inWholeMicroseconds / 1000.0

			if (round == 0) {
				first = build
			}
			if (round >= WARMUP_ROUNDS) {
				val times = doubleArrayOf(build, rank, parse)
				for (i in times.indices) {
					if (times[i] < best[i]) {
						best[i] = times[i]
					}
				}
				ranked = r
				places = p
			}
		}
		regions.close()
		println("")
		println("### search words, shared copy on ${testPlatformName()}, best of $MEASURED_ROUNDS")
		println("  ${"what".padEnd(34)} ${"ms".padStart(9)} ${"count".padStart(9)}")
		row("first build in the process", first, 1)
		row("build the words", best[0], 1)
		row("rank ${words.size} words", best[1], ranked)
		row("parse ${phrases.size} phrases", best[2], places)
	}

	private fun row(what: String, ms: Double, count: Int) {
		val rounded = (kotlin.math.round(ms * 10) / 10).toString()
		println("  ${what.padEnd(34)} ${rounded.padStart(9)} ${count.toString().padStart(9)}")
	}

	companion object {
		const val WARMUP_ROUNDS = 3
		const val MEASURED_ROUNDS = 5
	}
}
